tools/generate_schema_wrapper.py

```diff
@@ -859,48 +863,64 @@ def vegalite_main(skip_download: bool = False) -> None:
         ruff_write_lint_format_str(fp, contents)
 
 
-def _create_encode_signature(
-    channel_infos: dict[str, ChannelInfo],
-) -> str:
-    signature_args: list[str] = ["self", "*args: Any"]
-    docstring_parameters: list[str] = ["", "Parameters", "----------"]
-
-    # TODO: refactor so extracting the `TypedDict` here makes sense
-    # - Want to avoid simply returning tuple[str, str]
-    typed_dict_args: list[str] = []
-
-    for channel, info in channel_infos.items():
-        # `Map` stands for the return types of alt.datum, alt.value as well as
-        # the dictionary representation of an encoding channel class. See
-        # discussions in https://github.com/vega/altair/pull/3208
-        # for more background.
-        union_types = ["str", info.field_class_name, "Map"]
-        it_rst_names = (rst_syntax_for_class(c) for c in info.all_names)
-        docstring_union_types = ["str", next(it_rst_names), "Dict"]
-        tp_inner: str = " | ".join(chain(union_types, info.non_field_names))
-        if info.supports_arrays:
-            # We could be more specific about what types are accepted in the list
-            # but then the signatures would get rather long and less useful
-            # to a user when it shows up in their IDE.
-            docstring_union_types.append("List")
-            tp_inner = f"OneOrSeq[{tp_inner}]"
-
-        signature_args.append(f"{channel}: Optional[{tp_inner}] = Undefined")
-        typed_dict_args.append(f"{channel}: {tp_inner}")
-        docstring_parameters.extend(
-            (
-                f"{channel} : {', '.join(chain(docstring_union_types, it_rst_names))}",
-                f"    {process_description(info.deep_description)}",
+@dataclass
+class EncodingArtifacts:
+    """
+    Wrapper for what was previously `_create_encode_signature()`.
+
+    Now also creates a `TypedDict` as part of https://github.com/pola-rs/polars/pull/17995
+    """
+
+    channel_infos: dict[str, ChannelInfo]
+    fmt_method: str
+    fmt_typed_dict: str
+
+    def __iter__(self) -> Iterator[str]:
```

Comment:
I enjoyed reading an iterable data class but I'm not sure if there is any advantage of writing it as such? It's used as something which takes 3 arguments and outputs a sequence -> Could be a simple function `generate_encoding_artifacts(channel_infos, fmt_method, fmt_typed_dict)` which would be more in line with the rest of the codebase. Just to keep the cognitive load as low as possible for new joiners to the codebase :)

Reply:
> I enjoyed reading an iterable data class but I'm not sure if there is any advantage of writing it as such?

I will admit it is the ugly duckling at the moment, no major objections if changed to a function (especially for something outside of `altair`).

>  Just to keep the cognitive load as low as possible for new joiners to the codebase :)

Following this up with another comment explaining the choice, since I think we're both seeking the same goal - but approaching it differently.
In case that takes a while to write up nicely, this would be one step *among others* to gradually break down a lot of the code into smaller pieces

Reply:
I took the liberty to rewrite it in [Rewrite EncodeArtifacts dataclass as a function](https://github.com/vega/altair/pull/3515/commits/79f317dcaa8b74fff6891530dab7b185fec489b5) but we can also revert it back.

Reply:
Just for context, your comment didn't load for me before I wrote mine :)

Reply:
> I took the liberty to rewrite it in [Rewrite EncodeArtifacts dataclass as a function](https://github.com/vega/altair/pull/3515/commits/79f317dcaa8b74fff6891530dab7b185fec489b5) but we can also revert it back.

As a compromise, could we do:

```py
def generate_encoding_artifacts(
    channel_infos: dict[str, ChannelInfo], fmt_method: str, fmt_typed_dict: str
) -> Iterator[str]:
    # do stuff
    # ...
	artifacts = ["1", "2"]
    yield from artifacts
```

Essentially then you can update `generate_encoding_artifacts` to do whatever in the future, but keep the same annotation and usage here:

```py
*generate_encoding_artifacts(channel_infos, ENCODE_METHOD, ENCODE_TYPED_DICT)
```

So this always means the same thing, regardless of how many `artifacts`

Reply:
Changed it in https://github.com/vega/altair/pull/3515/commits/0287eba43a18b8a4b4a4c2963e50c92f23726bc5 to an iterator as suggested. Feels like I can learn something here :) To put it into different words to see if I understood it: Is it correct that you proposed this so that there is no promise that `generate_encoding_artifacts` always returns just 2 strings but could be expanded in the future?

Reply:
> Changed it in [0287eba](https://github.com/vega/altair/commit/0287eba43a18b8a4b4a4c2963e50c92f23726bc5) to an iterator as suggested. Feels like I can learn something here :) To put it into different words to see if I understood it: Is it correct that you proposed this so that there is no promise that `generate_encoding_artifacts` always returns just 2 strings but could be expanded in the future?

Yeah exactly you've got it!


I'm not sure if you've run into this before, but for me when a function returns multiple things - I then need to think *"what is each one?"*, *"does the order matter?"*, etc.

Now it is just that this returns strings and it does that exactly once

Reply:
Definitely! Great if code can make these things clearer. Will also use it more going forward, thanks.

Reply:
Some more useful examples `alt.when` following [api._parse_when_compose](https://github.com/vega/altair/blob/95f654fcd036069e1101883d176ec67c6f13fa25/altair/vegalite/v5/api.py#L711)